### PR TITLE
Also draw slope annotation with gradient=0 (#10541)

### LINKED ISSUE #10541: [BUG]Slope annotation is not drawn with gradient of 0
#### ALL software version info (bokeh, python, notebook, OS, browser, any other relevant packages)
python: 3.8.5
bokeh: 2.2.1
jupyter_core: 4.6.3
jupyter_client: 6.1.6 
windows 10 Home

#### Description of expected behavior and the observed behavior
Slope does not draw any line when `gradient == 0`. I would expect it to draw a horizontal line. Seem funky to have to check if my gradient is near 0 and have an if statement to draw a horizontal Span instead of being able to specify `Slope(gradient=0, y_intercept=...)`

#### Complete, minimal, self-contained example code that reproduces the issue
```
# code goes here between backticks
from bokeh.plotting import figure
from bokeh.models import Slope
from bokeh.io import show

p = figure(width=200, height=200, x_range=(-1, 1), y_range=(-1, 1))

for gradient in [1, -1, 0]:
    s = Slope(gradient=gradient, y_intercept=0)
    p.add_layout(s)

show(p)
```

#### Stack traceback and/or browser JavaScript console output
No errors or warnings are raised

#### Screenshots or screencasts of the bug in action
Actual output:
image

Expected output:
image

### PR DESCRIPTION
**Pull request by harmbuisman**
Fixes #10541

A horizontal line is drawn for a slope with gradient=0 at the level of the y_intersept.

Tested using the following:
```
from bokeh.plotting import figure, output_file
from bokeh.models import Slope
from bokeh.io import show
from bokeh.io import curdoc

p = figure(width=200, height=200, x_range=(-5, 5), y_range=(-5, 5))

for gradient in [1, -1, 0, 2, -0.5]:
    s = Slope(gradient=gradient, y_intercept=-1)
    p.add_layout(s)

curdoc().add_root(p)
```

afbeelding

### CODE REVIEW

**Comment by mattpap** [comment]
This needs to be visually tested under `bokehjs/test/integration/regressions.ts`. See e.g. https://github.com/bokeh/bokeh/blob/dcaeba25b720528e32460a988a364733389fcd5f/bokehjs/test/integration/regressions.ts#L229-L243 for inspiration. Use your provided example for the test (include a vertical line as well).

**Comment by mattpap** [comment]
Rebase on top of `branch-2.3` to resolve the unrelated codebase test suite failure.

File: bokehjs/src/lib/models/annotations/slope.ts [hunk 1]
```diff
@@ -29,17 +29,29 @@ export class SlopeView extends AnnotationView {
...
+      const y_end = yscale.invert(sy_end)
 
-    const sx_start = xscale.compute(x_start)
-    const sx_end = xscale.compute(x_end)
+      const x_start = (y_start - y_intercept) / gradient
+      const x_end = (y_end - y_intercept) / gradient
+
+      sx_start = xscale.compute(x_start)
+      sx_end = xscale.compute(x_end)
+    }
```
Comment (bryevdv) [inline bokehjs/src/lib/models/annotations/slope.ts:54]: This method is getting a bit logn. Would suggest pulling the `sx_start`, `sy_end`, etc code in to a helper method. Then you can also avoid the `let` statements by using early return:
```
if (gradient == 0) {
  // compute zero case
  return [...]
}
// compute non-zero case
return [...]
```

**Comment by harmbuisman** [comment]
@mattpap 

Thanks. The vertical line is not defined for slope, as it would have infinite slope and non-existing intercept.

For the test I do get stuck. I tried the following: 

 ```
 describe("in issue #10541", () => { 
    it("prevents slope with gradient=0", async () => { 

      const p = fig([200, 200], {x_range:[-5, 5], y_range:[-5, 5]})

      for (const gradient in [1, -1, 0, 2, -0.5]){
        p.slope({gradient:gradient, y_intercept:-1})
      }
   
      await display(p, [350, 350]) 
    }) 
  }) 
```

But p.slope does not exist. I can only find the Python way of declaring a slope. Where should I look the typescript ways of declaring it? Would the loop work?

How can I run this test to validate if it is implemented correctly?

**Comment by mattpap** [comment]
> But p.slope does not exist. I can only find the Python way of declaring a slope.

The model API is more or less the same in bokehjs, so this:
```ts
const s = new Slope({gradient, y_intercept: -1})
p.add_layout(s)
```
should work.

> The vertical line is not defined for slope, as it would have infinite slope and non-existing intercept.

Right. Ignore that.

File: bokehjs/src/lib/models/annotations/slope.ts (hunk 1 above)
Comment (harmbuisman) [reply bokehjs/src/lib/models/annotations/slope.ts:51]: @bryevdv, thanks for your suggestion. I tried a few things, but I lack the fluency in typescript to get the scoping right without VSCode warnings and without using a lot of code. Any pointer is welcome. I now reduced it by three lines by declaring the variables in one as I saw also done in the arrow.ts in the same directory.

**Comment by mattpap** [comment]
> `for (const gradient in [1, -1, 0, 2, -0.5]) {`

Should be `for of`. With the about you're iterating over indices of an array.

>  `await display(p, [350, 350])`

Make it 250 by 250.

**Comment by harmbuisman** [comment]
@mattpap, I added the test, but I am having trouble generating the baseline image

I get an error that the --set-new-base-screenshot does not exist. Any pointer?
afbeelding

**Comment by mattpap** [comment]
You should follow https://docs.bokeh.org/en/latest/docs/dev_guide/bokehjs.html#ci-and-visual-testing, or in short:
1. `cd bokehjs; node make test`
2. commit two generated baseline files
3. push
4. download and unzip the artifact
5. commit baselines for other platforms
6. push -f

File: bokehjs/test/integration/regressions.ts [hunk 2]
```diff
@@ -9,7 +9,7 @@ import {
   GlyphRenderer, GraphRenderer, GridBox,
   Circle, Quad, MultiLine,
   StaticLayoutProvider,
-  Plot,
+  Plot, Slope
```
Comment (mattpap) [inline bokehjs/test/integration/regressions.ts:12]: Trailing comma.

File: bokehjs/test/integration/regressions.ts [hunk 3]
```diff
@@ -591,6 +591,20 @@ describe("Bug", () => {
     })
   })
 
+  describe("in issue #10541", () => {
+    it("prevents Slope with gradient=0", async () => {
+
+      const p = fig([200, 200], {x_range:[-5, 5], y_range:[-5, 5]})
+
+      for (const gradient of [1, -1, 0, 2, -0.5]){
+        const s = new Slope({gradient:gradient, y_intercept:-1})
```
Comment (mattpap) [inline bokehjs/test/integration/regressions.ts:600]: Use shorthand syntax, i.e. `{gradient, y_intercept: -1}`.

**Pull request merged**